File: typescript/tests/text_embedder.test.ts
// text_embedder.test.ts
//
// Exercises CircleAI.Embeddings.TextEmbedder with an injected model manager +
// backend factory (no native MNN library). Verifies the resolve → verify →
// build → embed handshake and l2Normalize.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  TextEmbedder,
  l2Normalize,
  type IEmbeddingBackend,
  type EmbeddingBackendFactory,
} from '../src/embeddings/index';
import type { IModelManager } from '../src/core/index';

/** Fake model manager: records the verify calls and its verdict is toggleable. */
class FakeModelManager implements IModelManager {
  resolved = 0;
  verified = 0;
  constructor(
    private readonly modelPath: string,
    private readonly verdict: boolean,
  ) {}
  getModelPathAsync(_modelId: string): Promise<string> {
    this.resolved++;
    return Promise.resolve(this.modelPath);
  }
  verifyModelAsync(_modelPath: string, _expected: Uint8Array): Promise<boolean> {
    this.verified++;
    return Promise.resolve(this.verdict);
  }
  dispose(): void {}
}

/** Backend that returns a fixed (already-normalised) vector and counts builds. */
function fakeBackend(dim: number): { factory: EmbeddingBackendFactory; builds: () => number } {
  let builds = 0;
  const factory: EmbeddingBackendFactory = (_path: string): IEmbeddingBackend => {
    builds++;
    return {
      dimension: dim,
      embed: (text: string) => {
        const v = new Float32Array(dim);
        for (let i = 0; i < dim; i++) v[i] = (text.length + i) % 7;
        l2Normalize(v);
        return v;
      },
      dispose: () => {},
    };
  };
  return { factory, builds: () => builds };
}

describe('l2Normalize', () => {
  it('normalises to unit length', () => {
    const v = Float32Array.from([3, 4]);
    l2Normalize(v);
    const norm = Math.hypot(v[0], v[1]);
    assert.ok(Math.abs(norm - 1) < 1e-6);
  });

  it('leaves a zero vector unchanged', () => {
    const v = new Float32Array([0, 0, 0]);
    l2Normalize(v);
    assert.deepEqual(Array.from(v), [0, 0, 0]);
  });
});

describe('TextEmbedder', () => {
  it('resolves + verifies the model once, then embeds', async () => {
    const mgr = new FakeModelManager('/models/emb', true);
    const be = fakeBackend(8);
    const embedder = new TextEmbedder(mgr, new Uint8Array([1, 2, 3]), be.factory);

    const a = await embedder.generateAsync('hello');
    const b = await embedder.generateAsync('world!');
    assert.equal(a.length, 8);
    assert.equal(b.length, 8);
    // Lazy init happens exactly once even across calls.
    assert.equal(mgr.resolved, 1);
    assert.equal(mgr.verified, 1);
    assert.equal(be.builds(), 1);
  });

  it('shares a single initialisation across concurrent callers', async () => {
    const mgr = new FakeModelManager('/models/emb', true);
    const be = fakeBackend(4);
    const embedder = new TextEmbedder(mgr, new Uint8Array([9]), be.factory);
    await Promise.all([
      embedder.generateAsync('a'),
      embedder.generateAsync('b'),
      embedder.generateAsync('c'),
    ]);
    assert.equal(be.builds(), 1);
    assert.equal(mgr.resolved, 1);
  });

  it('throws when checksum verification fails', async () => {
    const mgr = new FakeModelManager('/models/emb', false);
    const be = fakeBackend(4);
    const embedder = new TextEmbedder(mgr, new Uint8Array([1]), be.factory);
    await assert.rejects(
      () => embedder.generateAsync('x'),
      /checksum verification failed/,
    );
    // Failed init is not cached — a later call retries the handshake.
    assert.equal(be.builds(), 0);
  });

  it('rejects empty text', async () => {
    const mgr = new FakeModelManager('/models/emb', true);
    const embedder = new TextEmbedder(mgr, new Uint8Array([1]), fakeBackend(4).factory);
    await assert.rejects(() => embedder.generateAsync(''), /cannot be empty/);
    await assert.rejects(() => embedder.generateAsync('   '), /cannot be empty/);
  });

  it('throws after dispose', async () => {
    const mgr = new FakeModelManager('/models/emb', true);
    const embedder = new TextEmbedder(mgr, new Uint8Array([1]), fakeBackend(4).factory);
    await embedder.generateAsync('warm');
    embedder.dispose();
    await assert.rejects(() => embedder.generateAsync('x'), /disposed/);
  });

  it('validates constructor arguments', () => {
    const mgr = new FakeModelManager('/m', true);
    assert.throws(
      () => new TextEmbedder(null as unknown as IModelManager, new Uint8Array(), fakeBackend(4).factory),
      /modelManager is required/,
    );
    assert.throws(
      () => new TextEmbedder(mgr, null as unknown as Uint8Array, fakeBackend(4).factory),
      /expectedChecksum is required/,
    );
    assert.throws(
      () => new TextEmbedder(mgr, new Uint8Array(), null as unknown as EmbeddingBackendFactory),
      /backendFactory is required/,
    );
  });
});

File: rust/tests/text_embedder_test.rs
//! text_embedder_test.rs
//!
//! Verifies CircleAI.Embeddings.TextEmbedder port: model resolve + verify gating,
//! empty-text rejection, deterministic hashing backend, and L2-normalised output.

use std::path::{Path, PathBuf};

use circle_ai::embeddings::{
    EmbedderError, HashingEmbeddingBackend, IEmbeddingBackend, ITextEmbedder, TextEmbedder,
};
use circle_ai::model_runtime::{IModelManager, ModelLoaderError};

/// Mock manager: returns a fixed path and a caller-chosen verification verdict.
struct MockManager {
    verify_result: bool,
}
impl IModelManager for MockManager {
    fn get_model_path(&self, _model_id: &str) -> Result<PathBuf, ModelLoaderError> {
        Ok(PathBuf::from("embedding/model.gguf"))
    }
    fn verify_model(
        &self,
        _model_path: &Path,
        _expected_checksum: &[u8],
    ) -> Result<bool, ModelLoaderError> {
        Ok(self.verify_result)
    }
}

fn l2(v: &[f32]) -> f32 {
    v.iter().map(|&x| x as f64 * x as f64).sum::<f64>().sqrt() as f32
}

#[test]
fn hashing_backend_is_deterministic_and_unit_norm() {
    let b = HashingEmbeddingBackend::new(64);
    assert_eq!(b.dimension(), 64);
    let a = b.embed("the quick brown fox").unwrap();
    let c = b.embed("the quick brown fox").unwrap();
    assert_eq!(a, c);
    assert_eq!(a.len(), 64);
    assert!((l2(&a) - 1.0).abs() < 1e-5, "not unit norm: {}", l2(&a));
}

#[test]
fn hashing_backend_distinguishes_texts() {
    let b = HashingEmbeddingBackend::new(128);
    let a = b.embed("apples").unwrap();
    let c = b.embed("oranges").unwrap();
    assert_ne!(a, c);
}

#[test]
fn embedder_generates_unit_vector() {
    let mgr = MockManager { verify_result: true };
    let embedder = TextEmbedder::new(mgr, vec![1, 2, 3]);
    let v = embedder.generate("hello world").unwrap();
    assert_eq!(v.len(), 384);
    assert!((l2(&v) - 1.0).abs() < 1e-5);
}

#[test]
fn embedder_rejects_empty_text() {
    let mgr = MockManager { verify_result: true };
    let embedder = TextEmbedder::new(mgr, vec![1]);
    assert!(matches!(
        embedder.generate("   "),
        Err(EmbedderError::Argument(_))
    ));
}

#[test]
fn embedder_fails_when_checksum_unverified() {
    let mgr = MockManager {
        verify_result: false,
    };
    let embedder = TextEmbedder::new(mgr, vec![9, 9]);
    assert!(matches!(
        embedder.generate("anything"),
        Err(EmbedderError::InvalidData(_))
    ));
}

#[test]
fn embedder_uses_injected_backend_factory() {
    // Inject a backend factory producing a fixed-dim backend and confirm the
    // resolved model path is threaded through.
    let mgr = MockManager { verify_result: true };
    let embedder = TextEmbedder::with_backend_factory(
        mgr,
        vec![0],
        Box::new(|path: &str| {
            assert!(path.contains("model.gguf"));
            Box::new(HashingEmbeddingBackend::new(16)) as Box<dyn IEmbeddingBackend>
        }),
    );
    let v = embedder.generate("token one two").unwrap();
    assert_eq!(v.len(), 16);
}

#[test]
fn embedder_initialises_backend_once() {
    // Two calls reuse the same backend (init gate); both succeed and match.
    let mgr = MockManager { verify_result: true };
    let embedder = TextEmbedder::new(mgr, vec![1]);
    let a = embedder.generate("stable text").unwrap();
    let b = embedder.generate("stable text").unwrap();
    assert_eq!(a, b);
}
